File: C++/1088.cpp
#include <iostream>
using namespace std;


bool canRook(int x1, int y1, int x2, int y2)
{
    return x1 == x2 || y1 == y2;
}


bool canBishop(int x1, int y1, int x2, int y2)
{
    int dx = x1 - x2;
    int dy = y1 - y2;

    if (dx < 0)
        dx = -dx;
    if (dy < 0)
        dy = -dy;

    return dx == dy;
}


bool canKnight(int x1, int y1, int x2, int y2)
{
    int dx = x1 - x2;
    int dy = y1 - y2;

    if (dx < 0)
        dx = -dx;
    if (dy < 0)
        dy = -dy;

    return (dx == 2 && dy == 1) || (dx == 1 && dy == 2);
}


bool canQueen(int x1, int y1, int x2, int y2)
{
    int dx = x1 - x2;
    int dy = y1 - y2;

    if (dx < 0)
        dx = -dx;
    if (dy < 0)
        dy = -dy;

    return x1 == x2 || y1 == y2 || dx == dy;
}


bool canKing(int x1, int y1, int x2, int y2)
{
    int dx = x1 - x2;
    int dy = y1 - y2;

    if (dx < 0)
        dx = -dx;
    if (dy < 0)
        dy = -dy;

    return (x1 == x2 && dy == 1) || (y1 == y2 && dx == 1)
                                || (dx == dy && dx == 1);
}


bool canPawn(int x1, int y1, int x2, int y2)
{
    int dx = x2 - x1;
    int dy = y2 - y1;

    return y1 > 1 && dx == 0 && (dy == 1 || (dy == 2 && y1 == 2));
}


int main()
{
    char coord1[3], coord2[3];
    int x1, y1, x2, y2;
    bool hasFigure = false;

    cin >> coord1 >> coord2;

    x1 = coord1[0] - 'A' + 1;
    y1 = coord1[1] - '0';
    x2 = coord2[0] - 'A' + 1;
    y2 = coord2[1] - '0';

    if (canRook(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "Rook" << endl;
    }
    if (canBishop(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "Bishop" << endl;
    }
    if (canKnight(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "Knight" << endl;
    }
    if (canQueen(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "Queen" << endl;
    }
    if (canKing(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "King" << endl;
    }
    if (canPawn(x1, y1, x2, y2))
    {
        hasFigure = true;
        cout << "Pawn" << endl;
    }
    if (!hasFigure)
        cout << "Nobody" << endl;

    return 0;
}

File: Python/1088.py
def canRook(x1, y1, x2, y2):
    return x1 == x2 or y1 == y2


def canBishop(x1, y1, x2, y2):
    dx = abs(x1 - x2)
    dy = abs(y1 - y2)
    return dx == dy


def canKnight(x1, y1, x2, y2):
    dx = abs(x1 - x2)
    dy = abs(y1 - y2)
    return dx == 2 and dy == 1 or dx == 1 and dy == 2


def canQueen(x1, y1, x2, y2):
    dx = abs(x1 - x2)
    dy = abs(y1 - y2)
    return x1 == x2 or y1 == y2 or dx == dy


def canKing(x1, y1, x2, y2):
    dx = abs(x1 - x2)
    dy = abs(y1 - y2)
    return x1 == x2 and dy == 1 or y1 == y2 and dx == 1 or dx == dy and dx == 1


def canPawn(x1, y1, x2, y2):
    dx = x2 - x1
    dy = y2 - y1

    return  y1 > 1 and dx == 0 and (dy == 1 or dy == 2 and y1 == 2)


coord1, coord2 = input().split()

x1 = ord(coord1[0]) - ord('A') + 1
y1 = ord(coord1[1]) - ord('0')
x2 = ord(coord2[0]) - ord('A') + 1
y2 = ord(coord2[1]) - ord('0')

hasFigure = False
if canRook(x1, y1, x2, y2):
    hasFigure = True
    print('Rook')
if canBishop(x1, y1, x2, y2):
    hasFigure = True
    print('Bishop')
if canKnight(x1, y1, x2, y2):
    hasFigure = True
    print('Knight')
if canQueen(x1, y1, x2, y2):
    hasFigure = True
    print('Queen')
if canKing(x1, y1, x2, y2):
    hasFigure = True
    print('King')
if canPawn(x1, y1, x2, y2):
    hasFigure = True
    print('Pawn')
if not hasFigure:
    print('Nobody')
